pkg/controller/machine-set-boot-image/vsphere_helpers.go

```diff
@@ -0,0 +1,376 @@
+		// upload the vmdk to which ever host that was first
+		// available with the required network and datastore.
+		err = upload(ctx, archive, lease, i)
+		if err != nil {
+			return fmt.Errorf("failed to upload: %w", err)
+		}
+	}
+
+	err = lease.Complete(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to lease complete: %w", err)
+	}
+
+	vm := object.NewVirtualMachine(client.Client, info.Entity)
+	if vm == nil {
+		return fmt.Errorf("error VirtualMachine not found, managed object id: %s", info.Entity.Value)
+	}
+	if secureBoot {
+		bootOptions, err := vm.BootOptions(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to get boot options: %w", err)
+		}
+		bootOptions.EfiSecureBootEnabled = ptr.To(false)
+
+		err = vm.SetBootOptions(ctx, bootOptions)
+		if err != nil {
+			return fmt.Errorf("failed to set boot options: %w", err)
+		}
+	}
+
+	err = vm.MarkAsTemplate(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to mark vm as template: %w", err)
+	}
+
+	err = attachTag(ctx, tagManager, vm.Reference().Value, infraID)
+	if err != nil {
+		return fmt.Errorf("failed to attach tag: %w", err)
+	}
+
+	for _, tagID := range providerSpec.TagIDs {
+		err = attachTag(ctx, tagManager, vm.Reference().Value, tagID)
+		if err != nil {
+			return fmt.Errorf("failed to attach tag: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func getClientsFromServerURL(ctx context.Context, server string) (*govmomi.Client, *tags.Manager, error) {
+	vcenterURL, err := url.Parse(server)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	client, err := govmomi.NewClient(ctx, vcenterURL, true)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	restClient := rest.NewClient(client.Client)
+	err = restClient.Login(ctx, vcenterURL.User)
+	if err != nil {
+		logoutErr := client.Logout(context.TODO())
+		if logoutErr != nil {
+			err = logoutErr
+		}
+		return nil, nil, err
+	}
+
+	tagManager := tags.NewManager(restClient)
+
+	return client, tagManager, nil
+}
+
+func createNewVMTemplate(streamData *stream.Stream, providerSpec *machinev1beta1.VSphereMachineProviderSpec, infra *osconfigv1.Infrastructure, arch string) (string, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Find and download the relevant OVA file
+	ova, err := streamData.QueryDisk(arch, "vmware", "ova")
+	if err != nil {
+		return "", err
+	}
+
+	klog.Infof("Downloading %s\n", ova.Location)
+	ovaPath, err := ova.Download(".")
+	if err != nil {
+		return "", fmt.Errorf("Failed to download %s: %w", ova.Location, err)
+	}
+	klog.Infof("Downloaded %s\n", ovaPath)
+
+	var name string
+	for _, vcenter := range infra.Spec.PlatformSpec.VSphere.VCenters {
+		if vcenter.Server != providerSpec.Workspace.Server {
+			continue
+		}
+
+		client, tagManager, err := getClientsFromServerURL(ctx, vcenter.Server)
+		if err != nil {
+			return "", err
+		}
+
+		finder := find.NewFinder(client.Client, false)
+
+		for _, failureDomain := range infra.Spec.PlatformSpec.VSphere.FailureDomains {
```

Comment:
Using just `Server` to determine the association between a failure domain and machineset is not enough.

Reply:
What would you recommend?

Reply:
updated